### DISCUSSION TITLE
problem with GL commands that run in pthread and they has dependency with each other

### DISCUSSION THREAD

**Discussion by aharondavid** [emscripten-core/emscripten#17417] [Q&A]
Hi, 
please have a look on the following pseudo code: 
it illustrate a real problem when you try to use GL in thread and main thread.
**any idea how to solve it will be welcome.**
```

Thread()
{
    //emscripten  post those commands  to the main thread by emscripten_sync_run_in_main_runtime_thread
    
    //drawing on canvas.
    glBindTexture(bind1);   
    glTexImage2D(tex1)
};

main_thread()
{
    createThread(Thread); //create a pthread

    //do somthing
   
    usleep(1);   //here, the busy wait mechanize is  executed

    //drawing on canvas.
    glBindTexture(bind2);
    glTexImage2D(tex2)
}
```

possible execution:
1. Thread run `glBindTexture(bind1)` , this command, internally, post and insert the glBindTexture to the main thread queue by emscripten_sync_run_in_main_runtime_thread
2. main thread  run `usleep `that do **busy wait**
3. **busy wait** is going to the main thread queue and exeute  `glBindTexture(bind1)` that was inserted to the queue by step 1.
4. main_thread  execute ` glBindTexture(bind2)`;
5. Thread run `glTexImage2D(tex1),` but it is work on bind2 insted of bind 1 !!!!

**Comment by sbc100** [comment]
Seems like this is general GL problem maybe? See https://stackoverflow.com/questions/22574879/using-the-same-opengl-context-in-multiple-threads

**Comment by sbc100** [comment]
Also: https://developer.apple.com/library/archive/documentation/GraphicsImaging/Conceptual/OpenGL-MacProgGuide/opengl_threading/opengl_threading.html#:~:text=OpenGL%20Restricts%20Each%20Context%20to%20a%20Single%20Thread,-Each%20thread%20in&text=If%20you%20modify%20the%20same,or%20it%20might%20render%20improperly.

"OpenGL is not reentrant. If you modify the same context from multiple threads simultaneously, the results are unpredictable. Your application might crash or it might render improperly."

**Comment by aharondavid** [comment]
the context are create for each thread by emscripten_webgl_create_context.
and the return context is well handled by emscripten code like in webgl1.c .
but, when I more think about that, I don't think that emscriptrn can do something to fix the above code.
